accounts/abi/unpack.go

```diff
@@ -113,15 +123,18 @@ func forEachUnpack(t Type, output []byte, start, size int) (interface{}, error)
 		return nil, fmt.Errorf("abi: invalid type in array/slice unpacking stage")
 	}
 
-	for i, j := start, 0; j*32 < len(slice); i, j = i+32, j+1 {
-		// this corrects the arrangement so that we get all the underlying array values
-		if t.Elem.T == ArrayTy && j != 0 {
-			i = start + t.Elem.Size*32*j
-		}
+	elemSize := 32
+	if t.T == ArrayTy {
+		elemSize = getDeepSizeForType(t.Elem)
+	}
```

Comment:
I would just write those 4 lines: `elemSize := getDeepSizeForType(t.Elem)` since `getDeppSizeForType` will return `32` if this is not an array.

Reply:
No, it doesn't. It is the element type that is being passed. So it returns 32 when *the element* is not array. Say you call `getDeepSizeForType(t.Elem)` with `t` being a slice, then we need the elemSize to be 32 bytes, since it's only the pointer to the contents of a slice-slot being counted, but `t.Elem` may be `ArrayTy`, resulting in something different than the 32 bytes.